src/util/json_encoder.c

```diff
@@ -0,0 +1,218 @@
+		s = _JsonEncoder_Edge((Edge *)ge, s);
+		break;
+	default:
+		ASSERT(false);
+	}
+	s = sdscat(s, "}");
+	return s;
+}
+
+static char *_JsonEncoder_Path(SIValue p, sds s) {
+	// open path with "["
+	s = sdscat(s, "[");
+
+	size_t nodeCount = SIPath_NodeCount(p);
+	for(size_t i = 0; i < nodeCount - 1; i ++) {
+		// write the next value
+		SIValue node = SIPath_GetNode(p, i);
+		s = _JsonEncoder_GraphEntity((GraphEntity *)&node, s, GETYPE_NODE);
+		s = sdscat(s, ", ");
+		SIValue edge = SIPath_GetRelationship(p, i);
+		s = _JsonEncoder_GraphEntity((GraphEntity *)&edge, s, GETYPE_EDGE);
+		s = sdscat(s, ", ");
+	}
+	// Handle last node.
+	if(nodeCount > 0) {
+		SIValue node = SIPath_GetNode(p, nodeCount - 1);
+		s = _JsonEncoder_GraphEntity((GraphEntity *)&node, s, GETYPE_NODE);
+	}
+
+	// close array with "]"
+	s = sdscat(s, "]");
+	return s;
+}
+
+static char *_JsonEncoder_Array(SIValue list, sds s) {
+	// open array with "["
+	s = sdscat(s, "[");
+	uint arrayLen = SIArray_Length(list);
+	for(uint i = 0; i < arrayLen; i ++) {
+		// write the next value
+		s = _JsonEncoder_SIValue(SIArray_Get(list, i), s);
+		// if it is not the last element, add ", "
+		if(i != arrayLen - 1) s = sdscat(s, ", ");
+	}
+
+	// close array with "]"
+	s = sdscat(s, "]");
+	return s;
+}
+
+static char *_JsonEncoder_Map(SIValue map, sds s) {
+	ASSERT(SI_TYPE(map) & T_MAP);
+
+	// "{" marks the beginning of a map
+	s = sdscat(s, "{");
+
+	uint key_count = Map_KeyCount(map);
+	for(uint i = 0; i < key_count; i ++) {
+		Pair p = map.map[i];
+		// write the next key/value pair
+		s = _JsonEncoder_String(p.key, s);
+		s = sdscat(s, ": ");
+		s = _JsonEncoder_SIValue(p.val, s);
+		// if this is not the last element, add ", "
+		if(i != key_count - 1) s = sdscat(s, ", ");
+	}
+
+	// "}" marks the end of a map
+	s = sdscat(s, "}");
+	return s;
+}
+
+sds _JsonEncoder_SIValue(SIValue v, sds s) {
+	switch(v.type) {
+	case T_STRING:
+		s = _JsonEncoder_String(v, s);
+		break;
+	case T_INT64:
+		s = sdscatfmt(s, "%I", v.longval);
+		break;
+	case T_BOOL:
+		if(v.longval) s = sdscat(s, "true");
+		else s = sdscat(s, "false");
+		break;
+	case T_DOUBLE:
+		s = sdscatprintf(s, "%f", v.doubleval);
+		break;
+	case T_NODE:
+		s = _JsonEncoder_GraphEntity(v.ptrval, s, GETYPE_NODE);
+		break;
+	case T_EDGE:
+		s = _JsonEncoder_GraphEntity(v.ptrval, s, GETYPE_EDGE);
+		break;
+	case T_ARRAY:
+		s = _JsonEncoder_Array(v, s);
+		break;
+	case T_MAP:
+		s = _JsonEncoder_Map(v, s);
+		break;
+	case T_PATH:
+		s = _JsonEncoder_Path(v, s);
+		break;
+	case T_NULL:
+		s = sdscat(s, "null");
+		break;
+	default:
+		// unrecognized type
+		printf("unrecognized type: %d\n", v.type);
+		ASSERT(false);
+		break;
+	}
+	return s;
+}
+
+char *JsonEncoder_SIValue(SIValue v) {
+	// Create an empty sds string.
+	sds s = sdsempty();
+	// Populate the sds string with encoded data.
+	s = _JsonEncoder_SIValue(v, s);
+	// Duplicate the sds string into a standard C string.
+	char *retval = rm_strdup(s);
```

Comment:
That's a bit of a shame, wish `sds` could have been freed without touching the actual string...

Reply:
Agreed, this is unfortunate. However, our string is `buf` in the struct:
```
struct __attribute__((__packed__)) sdshdr8 {
  uint8_t len;         /* used */
  uint8_t alloc;       /* excluding the header and null terminator */
  unsigned char flags; /* 3 lsb of type, 5 unused bits */
  char buf[];
};
```
I don't think there's any way to extricate the string from the header, and would prefer to not add `sds` logic to our `SIAllocation`.